src/tree-sitter-language-mode.js

```diff
@@ -947,55 +950,97 @@ class HighlightIterator {
       }
     );
 
-    this.iterators = [
-      this.languageMode.rootLanguageLayer.buildHighlightIterator()
-    ];
-    for (const marker of injectionMarkers) {
-      this.iterators.push(marker.languageLayer.buildHighlightIterator());
-    }
-    this.iterators.sort((a, b) => b.getIndex() - a.getIndex());
-
     const containingTags = [];
     const containingTagStartIndices = [];
     const targetIndex = this.languageMode.buffer.characterIndexForPosition(
       targetPosition
     );
-    for (let i = this.iterators.length - 1; i >= 0; i--) {
-      this.iterators[i].seek(
-        targetIndex,
-        containingTags,
-        containingTagStartIndices
-      );
+
+    this.iterators = [];
+    const iterator = this.languageMode.rootLanguageLayer.buildHighlightIterator();
+    if (iterator.seek(targetIndex, containingTags, containingTagStartIndices)) {
+      this.iterators.push(iterator);
     }
-    this.iterators.sort((a, b) => b.getIndex() - a.getIndex());
+    for (const marker of injectionMarkers) {
+      const iterator = marker.languageLayer.buildHighlightIterator();
+      if (
+        iterator.seek(targetIndex, containingTags, containingTagStartIndices)
+      ) {
+        this.iterators.push(iterator);
+      }
+    }
+    this.iterators.sort((a, b) => b.compare(a));
```

Comment:
I've also done a refactor which made it easier to implement the new functionality. I've changed the invariant of the `HighlightIterator` a bit. Previously, the the `iterators` array always contained one `HighlightLayerIterator` for *every* `HighlightLayer`. Layer iterators that were no longer active had a `done` flag set to `true`. Now, once a `HighlightLayerIterator` advances to the end of its syntax tree, I remove it from the `iterators` array. This removes the need for the `done` flag.

The reason for this diff is that `LayerIterator.seek` can now return `false`, which means that the iterator has already fallen off the end of its tree. If `seek` returns false, we avoid adding the iterator to the array at all.